openhtf/io/output/mfg_inspector.py

```diff
@@ -0,0 +1,276 @@
+# measurement.
+MAX_PARAMS_PER_MEASUREMENT = 100
+
+
+class UploadFailedError(Exception):
+  """Raised when an upload to mfg-inspector fails."""
+
+
+# pylint: disable=invalid-name
+def _PopulateHeader(record, testrun):
+  """Populate header-like info in testrun from record.
+
+  Mostly obvious, some stuff comes from metadata, see docstring of
+  TestRunFromTestRecord for details.
+  """
+  testrun.dut_serial = record.dut_id
+  testrun.tester_name = record.station_id
+  if 'test_name' in record.metadata:
+    testrun.test_info.name = record.metadata['test_name']
+  else:
+    # Default to copying tester_name into test_info.name.
+    testrun.test_info.name = record.station_id
+  if 'test_description' in record.metadata:
+    testrun.test_info.description = record.metadata['test_description']
+  if 'test_version' in record.metadata:
+    testrun.test_info.version_string = record.metadata['test_version']
+  testrun.test_status = OUTCOME_MAP.get(record.outcome, testrun_pb2.ERROR)
+  testrun.start_time_millis = record.start_time_millis
+  testrun.end_time_millis = record.end_time_millis
+  if 'run_name' in record.metadata:
+    testrun.run_name = record.metadata['run_name']
+  for details in record.outcome_details:
+    testrun_code = testrun.failure_codes.add()
+    if ':' in details:
+      testrun_code.code, testrun_code.details = details.split(':', 1)
+    else:
+      testrun_code.details = details
+
+
+def _AttachJson(record, testrun):
+  """Attach a copy of the JSON-ified record as an info parameter.
+
+  Save a copy of the JSON-ified record in an attachment so we can access
+  un-mangled fields later if we want.  Remove attachments since those get
+  copied over and can potentially be quite large.
+  """
+  record_dict = util.convert_to_dict(record, ignore_keys=('attachments',))
+  record_json = json_factory.OutputToJSON().encode(record_dict)
+  testrun_param = testrun.info_parameters.add()
+  testrun_param.name = 'OpenHTF_record.json'
+  testrun_param.value_binary = record_json
+  # pylint: disable=no-member
+  testrun_param.type = testrun_pb2.InformationParameter.TEXT_UTF8
+  # pylint: enable=no-member
+
+
+def _ExtractAttachments(phase, testrun, used_parameter_names):
+  """Extract attachments, just copy them over."""
+  for name, (data, mimetype) in phase.attachments.iteritems():
+    while name in used_parameter_names:
+      name += '_'  # Hack to avoid collisions between phases.
+    used_parameter_names.add(name)
+    testrun_param = testrun.info_parameters.add()
+    testrun_param.name = name
+    testrun_param.value_binary = data
+    if mimetype in MIMETYPE_MAP:
+      testrun_param.type = MIMETYPE_MAP[mimetype]
+    else:
+      # pylint: disable=no-member
+      testrun_param.type = testrun_pb2.InformationParameter.BINARY
+      # pylint: enable=no-member
+
+
+def _MangleMeasurement(name, value, measurement, mangled_parameters):
+  """Flatten parameters for backwards compatibility, watch for collisions.
```

Comment:
not in this change, but you're welcome to do it once this is merged, i don't know when i'm going to have the time to, but likely not for a while.